cmd/deploy/input_test.go

```diff
@@ -2,9 +2,279 @@ package main
+					Repository: "example",
+					Tag:        "bar",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: "nil",
+			wantErr:    "",
+		},
+		{
+			name: "basic fallback",
+			rawLookup: map[string]string{
+				"": "example",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+				},
+			},
+			wantDigest: "nil",
+			wantErr:    "",
+		},
+		{
+			name: "digest fallback",
+			rawLookup: map[string]string{
+				"": "example@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: "",
+			wantErr:    "",
+		},
+		{
+			name: "tag fallback",
+			rawLookup: map[string]string{
+				"": "example:tag",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Tag:        "tag",
+				},
+			},
+			wantDigest: "",
+			wantErr:    "",
+		},
+		{
+			name: "bad digest",
+			rawLookup: map[string]string{
+				"invalid digest": "example:foo",
+			},
+			wantErr: `lookup key "invalid digest" invalid:`,
+		},
+		{
+			name: "bad reference",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "://#",
+			},
+			wantErr: "failed to parse lookup ref",
+		},
+		{
+			name: "digest mistmatch",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo@sha256:00f42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+			},
+			wantErr: "digest on lookup ref should either be omitted or match key:",
+		},
+	} {
+		t.Run(x.name, func(t *testing.T) {
+			lookups, lookupDigest, err := normalizeInputLookup(x.rawLookup)
+
+			if x.wantErr != "" {
+				if err == nil {
+					t.Fatalf("Expected error not returned: %s", x.wantErr)
+				}
+
+				if !strings.Contains(err.Error(), x.wantErr) {
+					t.Fatalf("Expected error doesn't match.\ngot:\n%q,\n\nexpected to contain:\n%q", err, x.wantErr)
+				}
+				return
+			}
+
+			// This allows consistent display of <nil>
+			var wantDigest *digest.Digest
+			if x.wantDigest != "nil" {
+				v := digest.Digest(x.wantDigest)
+				wantDigest = &v
+			}
+			if (lookupDigest == nil && wantDigest != nil) ||
+				(lookupDigest != nil && wantDigest == nil) ||
+				(lookupDigest != nil && *lookupDigest != *wantDigest) {
+				t.Errorf("Unexpected digest.\ngot:\n%q\n\nexpected:\n%q", lookupDigest, wantDigest)
+			}
+
+			if !maps.Equal(lookups, x.wantLookup) {
+				t.Errorf("Lookups don't match.\ngot \n%s\n\nexpected\n%s", lookups, x.wantLookup)
```

Comment:
Is `%s` vs some permutation of `%v` intentional here?  It seems weird for a `map` to be printed via `%s` :see_no_evil:

(also `got \n` is weird -- is that trailing space intentional?)

Reply:
Huh! I just learned about `%#v`